accounts/api/views.py

```diff
@@ -25,4 +39,117 @@ class UserViewSet(viewsets.ModelViewSet):
+#   In this case, we need to write our own operations.
+# - Django rest framework url pattern: /resource/action/
+class AccountViewSet(viewsets.ViewSet):
+    serializer_class = SignupSerializer
+
+    # <Wayne Shih> 06-Aug-2021
+    # - This operation maps /accounts/login_status
+    # - detail refers to if this operation is on an specific resource obj
+    #   If true, then this operation maps /accounts/{user_id}/login_status
+    #   In this case, another argument pk needs to be passed in login_status()
+    # - methods refer to a list of http actions that allow
+    #   If an http action is not on the list, then it will return 405 - 'Method Not Allowed'
+    @action(methods=['GET'], detail=False)
+    def login_status(self, request):
+        """
+        Check current login status 
+        """
+        data = {'has_logged_in': request.user.is_authenticated}
+        if request.user.is_authenticated:
+            data['user'] = UserSerializer(request.user).data
+        return Response(data)
+
+    # <Wayne Shih> 06-Aug-2021
+    # Ref: https://docs.djangoproject.com/en/3.0/topics/auth/default/#django.contrib.auth.logout
+    @action(methods=['POST'], detail=False)
+    def logout(self, request):
+        """
+        Logout current user
+        """
+        django_logout(request)
+        return Response({'success': True})
+
+    # <Wayne Shih> 07-Aug-2021
+    # Ref:
+    # - https://www.django-rest-framework.org/api-guide/serializers/#field-level-validation
+    # - https://docs.djangoproject.com/en/3.2/topics/auth/default/#how-to-log-a-user-in
+    # - https://docs.djangoproject.com/en/3.2/ref/contrib/auth/#attributes
+    @action(methods=['POST'], detail=False)
+    def login(self, request):
+        """
+        Login
+        """
+        # <Wayne Shih> 07-Aug-2021
+        # get username and password from request
+        serializer = LoginSerializer(data=request.data)
+        if not serializer.is_valid():
+            return Response({
+                'success': False,
+                'message': 'Please check input.',
+                'errors': serializer.errors,
+            }, status=400)
+
+        # <Wayne Shih> 07-Aug-2021
+        # If validation is OK, then get user
+        username = serializer.validated_data['username']
+        password = serializer.validated_data['password']
+
+        if not User.objects.filter(username=username).exists():
+            # <Wayne Shih> 07-Aug-2021
+            # If you would like to be more secure, then just let django_authenticate()
+            # to handle this case and then response 'username and password does not match.'
+            # to avoid someone trying to get usernames.
+            # Try to print debug detail:
+            # print(User.objects.filter(username=username).query)
+            return Response({
+                'success': False,
+                'message': 'User does not exit.'
+            }, status=400)
```

Comment:
Is this possible to move this logic to serializer?
```suggestion


```

Reply:
Logic-wise, moving this block to serializer is doable.
The reason that I decided to make this logic here is that serializer is more like validating the input from request, 
but this check is more like validating the data in DB.
I would like to keep the check here.

As mentioned in the comment, in fact, another way we can choose is not to do this check at all and 
just make django_authenticate() to handle it.
In this case, if the user doesn't exist, django_authenticate() will return None and just show 
'username and password do not match.'
This way is more secure as well to avoid someone trying to get usernames.